### DISCUSSION TITLE
Redis 问题

### DISCUSSION THREAD

**Discussion by jxj807158342** [doocs/advanced-java#204] [Q&A]
数据更新 

1 .从微信接口获取  accesstoken 
2.删除 redis 缓存
3. 更新数据库 

今天  跟同事 讨论 他 说的是  先 第3步 再 2 第步，
如何反驳他，他说的理由是    高并发情况下 执行到第二步，   其他线程 查询，查询数据库的数据为旧数据

**Comment by teemozhu** [comment]
在有另外一个线程并发读取你设置的accesstoken的时候你同事说的是没有错误的，所以必须要双删，具体表现为
1. 从微信接口获取 accesstoken
2. 删除 redis 缓存
3. 更新数据库
4. 删除redis缓存

这个执行步骤在大部分的情况下是没有问题的

**Comment by huifer** [comment]
微信的 access_token 有一个过期时间, 如果是企业内的同一个微信access_token 应该要考虑多个组使用的情况. 
- 统一一个 access_token , 在一定时间内(7200秒) 整个与之相关的微信都只用这一个 access_token , 如果某一个应用重新获取了 access_token 那么其他也要重新获取. 此时不太友善吧. 

提出一个设计思路
1. 所有应用从一个地方获取 access_token . 对于这个 access_token 如果是redis 请设置过期时间, 将过期时间设置的比7200小一点(获取的时候可能就消费了一段时间), 7100 吧 嘿嘿嘿. 
2. 需要获取access_token的都从这个地方获取. 
3. 在第一步基础上开发一个定时任务, 每7100秒在重新获取数据 放入数据的方式如 [link](https://github.com/doocs/advanced-java/issues/204#issuecomment-730067147) 


不确定 access_token 是否是 7200 过期那个
